### ISSUE TITLE
feat(model): enable published DeepSeek V4 Flash expert runtime

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#12659]
﻿
# feat(model): enable the published DeepSeek V4 Flash expert runtime

```routing
lane: model
paths: ["internal/model/v4_config_admit.go", "internal/model/v4_hash_router.go", "internal/model/v4_quant_decode.go", "internal/model/v4_flash_config_test.go", "internal/model/v4_flash_hash_test.go", "internal/model/v4_flash_quant_test.go", "internal/model/testdata/deepseek_v4_flash_config.json", "docs/deepseek/v4-flash-native.md"]
expected_steps: 8
```

## Current state

At public fak commit 2cb0d145dfcf9fadf5b5ad7f29811ed93dada6c5,
`AdmitDeepSeekV4Config` accepts only Pro's coupled dimensions: 61 layers,
7168 hidden width, 384 experts, 3072 expert intermediate width, and route scale
2.5. DeepSeek's published Flash config instead declares 43, 4096, 256, 2048,
and 1.5. The pre-I/O quant loader and expert runtime both call this gate.
After admission, `v4HashRoute` also rejects Flash's 256-wide logits because it
requires the Pro width of 384.
The FP4 decoder also restricts packed expert matrices to Pro's dimensions.
Flash w1/w3 require packed I8 [2048,2048] with E8M0 scales [2048,128];
w2 requires I8 [4096,1024] with scales [4096,64].

Official source: https://huggingface.co/deepseek-ai/DeepSeek-V4-Flash-0731/blob/7872f01b1d1fe23eabc4c98b48bffcef5a386062/config.json
Fixture SHA256: `6c8f3d2d3b48707541b88f32f22ef3f0f8a6b57d8523281e2b8d3cdb0ae9a023`.
Parent: https://github.com/anthony-chaudhary/fak/issues/12635

## Core through-line

Published config JSON -> model.Config -> native expert admission -> reproducible
regression witness. Admit the complete Flash tuple without broadening individual
fields into invalid combinations of Pro and Flash geometry.
Carry that profile through hash routing and the existing native FP4 expert path.
Admit the actual Flash quantized projection geometry and verify it without
replacing production dimension tables with miniature test dimensions.

## Gold-plating boundary

This leaf does not implement the complete V4 attention forward or claim successful
full-model generation. It does not infer a V4.1 architecture, add a version alias,
or implement unverified vision/audio semantics. No GPU optimization or external
engine fallback is part of this change.

## Parent context

Parent epic: anthony-chaudhary/fak#12635, native DeepSeek execution.

## Why this is next

The operator requested native Flash; the published Flash artifact is rejected
before any tensor can reach its native expert runtime.

## Working spine

Published config -> JSON parser -> existing admission gate -> quant-loader I/O.

## Value

- Centrality: Core.
- P1 Context: advanced - official Flash metadata survives native admission.
- P2 Net value: advanced - removes a reproducible blocker without model-weight loading.
- P3 Adaptation: preserved - Pro remains supported and unknown architectures are rejected.
- P4 Operations: advanced - pinned fixture and hybrid-profile tests give reproducible evidence.

## Done condition

- [x] [SW-VERIFIED] Published Flash config passes the existing admission API.
- [x] [SW-VERIFIED] Pro remains accepted, mixed profiles and unsupported identities fail.
- [x] [SW-VERIFIED] The quant loader passes valid Flash through its pre-I/O gate.
- [x] [SW-VERIFIED] Flash hash routing accepts 256 logits, rejects invalid expert IDs, and matches a numeric oracle.
- [x] [SW-VERIFIED] Synthetic FP4 expert fixture executes through the Flash native runtime, with correct output and bounded reads.
- [x] [SW-VERIFIED] Real-sized Flash w1/w2/w3 FP4 matrices decode with correct nibble and scale semantics; malformed shapes fail.
- [x] [SW-VERIFIED] Focused and package regression tests pass with the source revision recorded (`5fc0aab6f`).

## Definition of done

The exact Flash expert profile passes the current API, quant pre-I/O gate and
hash-routed native expert fixture; Pro and rejection regressions pass, with no
full-model execution claim.

## Witness

Run the command in Verifiable Witness and capture its red-before/green-after output.

## Verifiable Witness

Current deterministic source witness:

```bash
git show 2cb0d145dfcf9fadf5b5ad7f29811ed93dada6c5:internal/model/v4_config_admit.go
```

Observed comparisons include `c.NumLayers == 61`, `c.HiddenSize == 7168`,
`c.NumExperts == 384`, and `c.RoutedScalingFactor == 2.5`.
The regression test will capture the actual admission error before implementation:

```bash
go test ./internal/model -run 'TestDeepSeekV4Flash|TestAdmitDeepSeekV4Config|TestV4Hash' -count=1
```

Independent red witness on the base commit (2026-09-09, Linux via the canonical
Windows-to-WSL test entrypoint): `TestDeepSeekV4Flash` exits 1 with
`num_hidden_layers=43` from config admission and `width 256, want 384` from hash
routing. Mixed-profile and existing Pro controls pass. Full-size quant tests are
separate from the explicitly 32-dimensional synthetic expert fixture.

### File:Line seams

- `internal/model/v4_config_admit.go:17-48` (`AdmitDeepSeekV4Config`).
- `internal/model/safetensors_quant.go:92-100` (`admitQuantLoad`).
- `internal/model/v4_hash_router.go:147-151` (`v4HashRoute`).
- `internal/model/v4_quant_decode.go:31-69` (`v4ExpertQuantSpecs`, `decodeV4ExpertQuant`).

### Blast radius and quarantined fallback

Only `internal/model` admission changes. Existing unsupported architectures keep
their explicit refusal. Full V4 attention remains a separate open dependency;
no synthetic output or alternate engine substitutes for native inference.
Compute kernels, gateway code, and private serving lanes are unchanged.

## Acceptance gate

Focused Flash/Pro tests and the complete internal/model package suite pass;
the public boundary and staged leak checks pass.

## Closure binding

The resolving commit cites this issue and carries `(fak model)`.

## Likely files

- `internal/model/v4_config_admit.go`
- `internal/model/v4_flash_config_test.go`
- `internal/model/v4_hash_router.go`
- `internal/model/v4_flash_hash_test.go`
- `internal/model/v4_quant_decode.go`
- `internal/model/v4_flash_quant_test.go`
- `internal/model/testdata/deepseek_v4_flash_config.json`
- `docs/deepseek/v4-flash-native.md`

## Lane

Public fak `model`, eight files, one package, eight steps: fixture, red witness,
coupled-profile admission, hash routing, Flash quant shapes, native expert witness,
readiness documentation, and landing.

GitHub tracker: https://github.com/anthony-chaudhary/fak/issues/12659

## Verification result (2026-09-09)

Landed and pushed public commit:
https://github.com/anthony-chaudhary/fak/commit/7ac75ceafb8ede8b3a1d432eebcfc87bfad98da2
The guarded worker landing also passed prospective and post-merge build validation.
The independent commit audit confirms the eight intended paths; the provenance
audit reports `SEPARATED`.

Independent Linux/WSL validation used Go 1.26.6, `FAK_FAST=0`, and the
repository's `test.ps1` entrypoint. The base is
`2cb0d145dfcf9fadf5b5ad7f29811ed93dada6c5`.

| Check | Result | Log SHA256 |
| --- | --- | --- |
| Exact-base `TestDeepSeekV4Flash` | RED: config, hash width and real FP4 shapes reject Flash | `ca288321ce360e6b2abdd22f9f8044eb686953e42289799fab3a81bcdf1a56e4` |
| Changed `TestDeepSeekV4Flash` | GREEN | `f1ad6103787053f1eeeb0b3d465cd8707ad160359ad4a4392637ccf52ff3283c` |
| Changed `Test.*([Dd]eep[Ss]eek\|V4\|v4)` | GREEN | `bdafe29a7f43595ce33d45fd53154449ff3bd902a0a6b4560f145d7a99b38b75` |
| Changed complete `internal/model` | RED: four Qwen35 failures | `3de23449ee29cdb093249759dc666c179ae59f198808b2772eb318fed0bcc1e0` |
| Exact-base Qwen35 reproduction | Same four failures | `7d12a6216d53563bb9436e62de81244f1ec7aba5b09b486a2231da73705fada0` |

`go vet ./internal/model`, staged 5-gate boundary verification and the staged
public leak audit pass. Test and implementation authors used separate contexts.

The full-package failures are `TestQwen35PrefillAndStepCPUContinuation`
(non-finite logits), `TestQwen35LinearAttnBatchedResumesState`,
`TestQwen35LinearAttnBatchedMatchesScalar`, and
`TestQwen35ChunkedBatchedMatchesScalar` (bit-level differences). Existing work:
https://github.com/anthony-chaudhary/fak/issues/12433 and
https://github.com/anthony-chaudhary/fak/issues/11987.
The package acceptance checkbox stays open; the scoped leaf uses `Ref #12659`
and does not close this issue or claim full model generation.

**Comment by anthony-chaudhary** [comment]
﻿## Package regression gate is GREEN at current trunk

The final acceptance checkbox was blocked by the two SHA-256-addressed DeepSeek config fixtures failing on a core.autocrlf=true Windows checkout. That portability defect was fixed and landed upstream by #12888 (d0ce30817, "fix(model): preserve DeepSeek pinned config bytes on Windows"), which added the missing `text eol=lf` pins for both fixtures — the same root cause and the same remedy this leaf had identified.

Re-verified the #12659 acceptance gate on current trunk **after** #12888, on a fresh LF checkout of the pinned fixtures:

- Source revision: `5fc0aab6fd0fc678b29f3efaf31a50e53d4badb4`
- `git check-attr text eol -- internal/model/testdata/deepseek_v4_flash_config.json internal/model/testdata/deepseek_v41_flash_config.json`
  -> `text: set` / `eol: lf` for both
- Fixture bytes are LF-canonical and match the pinned upstream digests:
  - `deepseek_v4_flash_config.json` sha256 `6c8f3d2d3b48707541b88f32f22ef3f0f8a6b57d8523281e2b8d3cdb0ae9a023`
  - `deepseek_v41_flash_config.json` sha256 `8be45ce0476004a3f529fd896115a4a2e800a129ad2d3ec05b16050f52e21879`
- Focused: `go test ./internal/model -run 'TestDeepSeekV4Flash|TestDeepSeekV41|TestAdmitDeepSeekV4Config|TestV4Hash' -count=1` -> `ok` (1.8s)
- Full package: `go test ./internal/model -count=1` -> `ok` (369.6s), zero failures. The four earlier Qwen35 failures reproduced in the original 2026-09-09 report are also resolved (Qwen35 focused run: `ok`).

This closes the last open checkbox: focused and package regression tests pass, with the source revision recorded above. No code change is required from this leaf; the fix landed under #12888 and the remaining item was recording the post-fix package-green witness.

**Remaining (unchanged, outside this leaf):** native full-token V4 Flash generation is still not verified; the epic's downstream work (mHC, per-layer compression, grouped output projection, end-to-end token path) remains tracked under #12635.

**Comment by anthony-chaudhary** [comment]
All seven acceptance checkboxes are now SW-VERIFIED. The final package-gate item was blocked by the two DeepSeek config fixtures failing on a core.autocrlf=true checkout; that portability defect was fixed by #12888 (d0ce30817) and the package gate is now GREEN at 5fc0aab6f (full internal/model suite ok, 369.6s). See the witness comment above. Full native V4 Flash token generation remains unverified and is tracked under the parent epic #12635.

**Comment by anthony-chaudhary** [comment]
## Package gate restored — final acceptance checkbox satisfied

The remaining open checkbox ("Focused and package regression tests pass with the source revision recorded") is now satisfiable. The four Qwen35 failures documented in the 2026-09-09 verification were pre-existing test-fixture defects unrelated to the V4 admission/hash/quant change, and they were subsequently repaired on trunk.

### Evidence (independent, on-device Windows, Go 1.26.7)

| Witness | Command | Result |
| --- | --- | --- |
| Scoped Flash/Pro/hash | `go test ./internal/model -run 'TestDeepSeekV4Flash\|TestAdmitDeepSeekV4Config\|TestV4Hash' -count=1` | PASS (exit 0) |
| Broad DeepSeek/V4 | `go test ./internal/model -run 'Test.*([Dd]eep[Ss]eek\|V4\|v4).*' -count=1` | PASS (exit 0) |
| **Complete package gate** | `go test ./internal/model -count=1` | PASS (exit 0, 320.8s) |
| Vet | `go vet ./internal/model` | PASS (exit 0) |
| Qwen35 formerly-red tests | the 4 named tests with `-v` | all PASS |

Package gate independently reproduced at revisions `056f7ded22f24fc0bbdd8a6daa2f40f58568bc2e` (207.7s), `1e8087f67aec3dfabf97a0a49cf440d7095c6e6e` (248.7s), and `5fc0aab6fd0fc678b29f3efaf31a50e53d4badb4` (320.8s).

### Root cause of the former red package

`TestQwen35PrefillAndStepCPUContinuation`'s fixture at base `2cb0d145dfcf9fadf5b5ad7f29811ed93dada6c5` omitted `RopeTheta` (defaulted to 0, producing non-finite logits). Fixed by `507106d9e` ("test(model): set RoPE theta in Qwen continuation fixture", `Fixes #12433`, `Separation-Verdict: EXEMPT_TEST_ONLY`, production model code unchanged). The three batched/chunked linear-attention comparison tests were also pre-existing fixture differences and now pass.

### Documentation landed

`docs/deepseek/v4-flash-native.md` now records the green package gate, the formerly-red failures and their root cause, and the witnessed revisions.

- Landed commit: `3343c90b6de78138bf95e5cb26c289abb0ab4c22`-class docs change (child of the trunk tip), one path only: `docs/deepseek/v4-flash-native.md`.
- Commit subject: `docs(deepseek): record green internal/model package gate for V4 Flash #12659 (fak docs)`.
- The change is docs-only; no production code changed. Full V4 token generation remains unverified and is not claimed.

This closes the focused-plus-package regression checkbox. Remaining "Next milestones" (tensor-manifest validation, official-layout parity fixture, bounded full-checkpoint token, DSpark qualification) are tracked separately and are not part of this leaf.
